### ISSUE TITLE
[data grid] Transpose columns into rows

### ISSUE THREAD

**Issue by jasonsturges** [mui/mui-x#3285]
### Summary 💡

Feature request to transpose the grid - reverse columns and rows.

Similar to pivot tables in #214, but just a matrix transform operation.

Changing from this:

| Header 1 | Header 2 |
| ------------- | ------------- |
| A 1 | B 1 |
| A 2 | B 2 |

...into this:

|  |  | |
| ------------- | ------------- | ------------- |
| **Header 1**  | A 1 | A 2 |
| **Header 2**  | B 1 | B 2 |

### Examples 🌈

Spreadsheets such as Microsoft Excel have a similar notion of transposing columns into rows, such as [here](https://support.microsoft.com/en-us/office/transpose-rotate-data-from-rows-to-columns-or-vice-versa-3419f2e3-beab-4318-aae5-d0f862209744) or [here](https://support.microsoft.com/en-us/office/transpose-function-ed039415-ed8a-4a81-93e9-4b6dfac76027)

![image](https://user-images.githubusercontent.com/1213591/143402661-599be1c0-8b7c-40fc-97c8-47da50dd94cc.png)

![image](https://user-images.githubusercontent.com/1213591/143402855-dff5318b-8325-4775-b467-fd94e60bde73.png)

### Motivation 🔦

Stems from a discussion in #3282 where I hoped to use editable data grids for property panels, such as in Unity:

![image](https://user-images.githubusercontent.com/1213591/143401080-75eb01a8-0251-4147-ad35-c22e7500fd41.png)

**Comment by jasonsturges** [comment]
@m4theushw Yes, I originally convoluted this together in #3282, but cleaned that issue up to isolate hiding headers.

Here, I was thinking either:
- Transpose / pivot / rotate table matrix to have columns as rows, which stacks headers in the first column
- Introduce the idea of Row Headers, like some grids / spreadsheets have

But you're right, this could be a two-column key / value pair grid - logic for that might be a bit more obtuse to handle, as there seemed to be an elegance if the data model was expressed via Data Grid columns.

If this feature request of transposing columns to rows, or introducing row headers doesn't resonate with the roadmap of Material UI Data Grids, I'll understand - this might simply be outside the intended usage here.

**Comment by m4theushw** [comment]
It's not in our roadmap, at least for now. I added the "waiting for 👍" label so other users with the same pain can upvote and we know if we should allocate time for it. The possibility of hiding the column headers is something we need to address.

**Comment by aimad-majdou** [comment]
Hello, any updates on this ?

**Comment by davidnir1** [comment]
Hello,
Was there any progress made on this?
Our users are driving us crazy asking for this :)

**Comment by cherniavskii** [comment]
Hey @aimad-majdou @davidnir1,
Could you please provide more details on your use case?
What is your data format and why would you want to transpose it?
Also, you might want to take a look at the pivoting issue https://github.com/mui/mui-x/issues/214 - maybe this is what you're looking for?
Thanks!

**Comment by shahamran** [comment]
Hi @cherniavskii, I can share my use case if it helps.

I have data with much more "columns" (headers) than "rows" (data points). Also, the header text can be quite long. Since text is arbitrarily wide, but only as high as the font size, it's more efficient (in terms of information density) to stack the headers vertically.

Consider this table:

| my very long column 1 | my very long column 2 | my very long column 3 | ... | my very long column N |
| --- | --- | --- | --- | --- |
| 0   | 1   | 2   | ... | N   |

this takes a lot of horizontal space, and will quickly need a scrollbar. But in this table:

|     |     |
| --- | --- |
| my very long column 1 | 1 |
| my very long column 2 | 2 |
| ... | ... |
| my very long column N | N |

the same information is displayed in a much more compact form. This remains true as long as there aren't many data points (in which case horizontal scrolling feels unnatural).

Hope this is clear :)

Also, I'm open for alternative ideas.

**Comment by jasonsturges** [comment]
Still interested in this - most simple example: pivot a column into a row in database query:

![pivot](https://github.com/mui/mui-x/assets/1213591/91c901e8-af6c-4cfd-9a51-d55489223763)

**Comment by stevstrong** [comment]
> Looking at the example you gave from Unity, couldn't your consider the param name as a column and the value as another column? It's basically reshaping the data to be rendered in a grid.

I would really appreciate if someone could give an example how to implement that.
Thanks in advance.

**Comment by MaximeDauphinot** [comment]
Hello, any news on it ? it would be awesome to possibly do this

**Comment by Marleypresidency65** [comment]
@davidnir1 https://amazon-elasticache834.github.io/uRyhe64Z3TvXTcAlnxRRNG9FtdM0NHF90sPHvV3WVUSSXhPVRFx41ziuD/
